Cython/Compiler/PyrexTypes.py

```diff
@@ -3776,32 +3776,62 @@ def can_coerce_from_pyobject(self, env):
             return True
         return False
 
+    def _get_conversion_context_cls_cname(self, cname_format_str):
+        # cname_format_str is passed "prefix", "format", "tags"
+        X = "XYZABC"
+        tags = []
+        context = {}
+        optional_tag_strs = []
+        optional_arg_template_names = ['']  # these may be different (e.g. for unordered_map, map)
+            # hence have to be generated here rather than hard-coded in the utility_code.
+            # Start with an empty string so that the joined output starts with ','
+        for ix, T in enumerate(self.templates or []):
+            if ix < builtin_cpp_conversions[self.cname]:
+                tags.append(T.specialization_name())
+                context[X[ix]] = T
+            else:
+                optional_arg_template_names.append("%s=*" % self.template_type.templates[ix].name)
+                if not isinstance(T, TemplatePlaceholderType):
+                    tags.append(T.specialization_name())
+                    optional_tag_strs.append(X[ix])
+                    context[X[ix]] = T
+
+        optional_tag_strs = ",".join(optional_tag_strs)
+        optional_arg_template_names = ",".join(optional_arg_template_names)
+
+        if self.cname in cpp_string_conversions:
+            cls = 'string'
+            prefix = 'PyObject_'  # gets specialised by explicit type casts in CoerceToPyTypeNode
+            tags = [type_identifier(self)]
+        else:
+            cls = self.cname[5:]
+            prefix = ''
+
+        cname = cname_format_str.format(prefix=prefix, cls=cls, tags='__and_'.join(tags))
+
+        context.update({
+            'cname': cname,
+            'maybe_unordered': self.maybe_unordered(),
+            'type': self.cname,
+            'optional_template_args': optional_tag_strs,
+            'optional_template_names': optional_arg_template_names,
+        })
+        return context, cls, cname
+
     def create_from_py_utility_code(self, env):
         if self.from_py_function is not None:
             return True
-        if self.cname in builtin_cpp_conversions or self.cname in cpp_string_conversions:
-            X = "XYZABC"
-            tags = []
-            context = {}
-            for ix, T in enumerate(self.templates or []):
-                if ix >= builtin_cpp_conversions[self.cname]:
-                    break
+        if self.cname in builtin_cpp_conversions:
+            # check that it's possible to convert all the relevant templates
+            # only scan the stored types; don't scan to hash types, comparison types, operators, etc.
+            for T in (self.templates or [])[:builtin_cpp_conversions[self.cname]]:
                 if T.is_pyobject or not T.create_from_py_utility_code(env):
```

Comment:
This condition seems slightly odd to me - T being an `object` shouldn't prevent conversion to/from Python.

However, since Cython doesn't handle C++ containers of `object` I think it's largely hypothetical so it's doing no harm